### PR TITLE
Ports TG's Garbage Subsystem

### PR DESCRIPTION
**Pull request by Fox-McCloud**
Ports over TG's garbage subsystem and all its various changes.

@MrStonedOne 

Also, implements `DROPDEL` which is a flag that, when applied to an object, makes it so it deletes itself when unequipped/dropped. No more having to do this:
```
/obj/item/thing/dropped()
	..()
	qdel(src)
```

### CODE REVIEW

File: code/__DEFINES/qdel.dm [hunk 1]
```diff
@@ -1,11 +1,31 @@
...
+#define GC_QUEUED_FOR_HARD_DEL -2
+#define GC_CURRENTLY_BEING_QDELETED -3
+
+#define QDELING(X) (X.gc_destroyed)
+#define QDELETED(X) (!X || QDELING(X))
+#define QDESTROYING(X) (!X || X.gc_destroyed == GC_CURRENTLY_BEING_QDELETED)
 
-// A convenient wrapper around checking for qdeletion/non-existence
-#define exists(A)	(A && !qdeleted(A))
+/proc/check_datum_qdeleted(datum/D) //for checking if something is a datum, first, then checking gc_destroyed; get rid of this eventually. TO-DO
```
Comment (Fox-McCloud) [inline code/__DEFINES/qdel.dm]: This is mostly for the damn timer system we have now; this will be deleted later on, once we port TG's timer SS.

File: code/__HELPERS/logging.dm [hunk 2]
```diff
@@ -15,9 +15,12 @@
 /proc/warning(msg)
 	log_to_dd("## WARNING: [msg]")
 
-//print a testing-mode debug message to world.log
-/proc/testing(msg)
```
Comment (Fox-McCloud) [inline code/__HELPERS/logging.dm]: @tigercat2000 

This is why we have weird MC logs at the moment.

File: code/controllers/subsystem/garbage.dm [hunk 3]
```diff
@@ -0,0 +1,436 @@
...
+		msg += "n/a|"
+	else
+		msg += "TGR:[round((totalgcs/(totaldels+totalgcs))*100, 0.01)]%"
+	msg += " P:[pass_counts.Join(",")]"
+	msg += "|F:[fail_counts.Join(",")]"
+	..(msg)
+
+/* TO-DO
+/datum/controller/subsystem/garbage/Shutdown()
+	//Adds the del() log to the qdel log file
```
Comment (Fox-McCloud) [inline code/controllers/subsystem/garbage.dm]: Leaving this here, just in case we want to have hard logs for this that persist out of round and want to plug this in.

File: code/__HELPERS/unsorted.dm [hunk 4]
```diff
@@ -1482,12 +1482,9 @@ var/mob/dview/dview_mob = new
 
 /mob/dview/New() //For whatever reason, if this isn't called, then BYOND will throw a type mismatch runtime when attempting to add this to the mobs list. -Fox
 
-/proc/IsValidSrc(var/A)
-	if(istype(A, /datum))
-		var/datum/B = A
-		return isnull(B.gcDestroyed)
-	if(istype(A, /client))
-		return 1
+/proc/IsValidSrc(datum/D)
```
Comment (tigercat2000) [inline code/__HELPERS/unsorted.dm]: You're going to have to readd clients to this or proc-calls in VV on /clients will not work.

File: code/modules/client/client defines.dm [hunk 5]
```diff
@@ -20,6 +20,13 @@
 
 	var/adminhelped = 0
 
+	var/gc_destroyed //Time when this object was destroyed.
```
Comment (Fox-McCloud) [inline code/modules/client/client defines.dm]: I'm guessing this isn't actually needed, but I included it just in case; we should probably make clients a child of `/datum` some time.

**Comment by Aurorablade** [comment]
Tesy it with cult datums i wanna see how much it explodes

**Comment by Fox-McCloud** [comment]
No more runtimes on startup.

Hopefully stupid cult datums stop causing havoc.

File: code/__HELPERS/global_lists.dm [hunk 6]
```diff
@@ -50,8 +50,6 @@
 
 	init_subtypes(/datum/crafting_recipe, crafting_recipes)
 
-	all_cults = typesof(/datum/cult_info)
```
Comment (Fox-McCloud) [inline code/__HELPERS/global_lists.dm]: Why?!

Comment (Aurorablade) [reply code/__HELPERS/global_lists.dm]: *well no one complained at the time...*

File: code/game/gamemodes/cult/ritual.dm [hunk 7]
```diff
@@ -42,13 +42,13 @@
 	if(!ticker.mode)
 		icon_state = "tome"
 	else
-		icon_state = ticker.mode.cultdat.tome_icon
+		icon_state = ticker.cultdat.tome_icon
```
Comment (Fox-McCloud) [inline code/game/gamemodes/cult/ritual.dm]: This should have been a damn proc, not spamming `ticker.cultdat.var_name_here` all over the damn place.

File: code/game/gamemodes/game_mode.dm [hunk 8]
```diff
@@ -374,7 +373,6 @@
 
 /datum/game_mode/New()
 	newscaster_announcements = pick(newscaster_standard_feeds)
-	cultdat = setupcult()
```
Comment (Fox-McCloud) [inline code/game/gamemodes/game_mode.dm]: Why. Why did this run every single time a new game-mode was created?

File: code/game/turfs/simulated/floor/plating.dm [hunk 9]
```diff
@@ -140,8 +140,8 @@
 
 /turf/simulated/floor/engine/cult/New()
 	..()
-	if(ticker.mode)//only do this if the round is going..otherwise..fucking asteroid..
-		icon_state = ticker.mode.cultdat.cult_floor_icon_state
+	if(ticker)//only do this if the round is going..otherwise..fucking asteroid..
```
Comment (Fox-McCloud) [inline code/game/turfs/simulated/floor/plating.dm]: say "we have a fire cult because the asteroid turfs are fire instead of normal" :|

File: code/game/gamemodes/game_mode.dm (hunk 8 above)
Comment (Aurorablade) [reply code/game/gamemodes/game_mode.dm]: at the time if it didn't..everything broke...cuase of how things loaded..like that cult ruin on the asteroid and loading icons.

File: code/game/gamemodes/cult/ritual.dm (hunk 7 above)
Comment (Aurorablade) [reply code/game/gamemodes/cult/ritual.dm]: my bad.

File: code/modules/admin/verbs/debug.dm [hunk 10]
```diff
@@ -748,6 +748,38 @@ But you can call procs that are of type /mob/living/carbon/human/proc/ for that
 		if("Respawnable Mobs")
 			to_chat(usr, jointext(respawnable_list,","))
 
+/client/proc/cmd_display_del_log()
```
Comment (Fox-McCloud) [inline code/modules/admin/verbs/debug.dm]: This one is insanely detailed, which is great, but sometimes you don't care about this level of input.

I'm tempted to code a secondary version that just lists things that failed, don't respect force, sleep in the destroy, and have no hint.

The old version was really easy to see at a glance without scrolling through every single thing deleted in the entire game, ever.

**Comment by Fox-McCloud** [comment]
Ready for review.

Reasonably confident for testing.

**Comment by DesolateG** [comment]
So can we get an explanation of what this does?

**Comment by Birdtalon** [comment]
Garbage collection queues things up to be deleted in a manageable way. The idea is to impact performance as little as possible.

**Comment by MrStonedOne** [comment]
the /tg/ version of the garbage subsystem delays the final reference removal until it fires next, so qdel() becomes a tiny bit faster because the actual soft delete doesn't happen until the next ssgarbage fire.

All in all it does a slightly better job of spreading out the cpu cost of garbage collection and management.

And with stonedmc, its a background subsystem so it never runs while the server is busy, only using spare cpu that would have gone to waste.

File: code/datums/datum.dm [hunk 11]
```diff
@@ -0,0 +1,14 @@
+/datum
```
Comment (tigercat2000) [inline code/datums/datum.dm]: I'm not sure I like adding any more files to the gross mess that is code/datums/
Any particular reason this can't just go in the GC code?

Comment (Fox-McCloud) [reply code/datums/datum.dm]: Parent types should generally get their own file and all their vars should be there; I'm not going to hunt down all the `/datum` vars and move them here in *this* PR though.

Comment (KasparoVy) [reply code/datums/datum.dm]: That's going to be a fun one.. 😱

Comment (MrStonedOne) [reply code/datums/datum.dm]: Ah the never ending tug of war between maintainability and modularity. Having the base level defines in their own files makes things more modular but having them in one file makes things easier to find.

Comment (Fox-McCloud) [reply code/datums/datum.dm]: I went ahead and moved the datum vars over to their correct file; wasn't nearly as many as I thought there would be.

Comment (tigercat2000) [reply code/datums/datum.dm]: Okay, let me rephrase: Why put this here instead of code/modules/datum?

Comment (Fox-McCloud) [reply code/datums/datum.dm]: This is where the vast majority of the other datums are stored that are generic and not tied with a particular module; typically you want the parent type kept close with the others--the base carbon/human/simple_animal/silicon types are defined close to where the other subtypes are.

The objects folder is structured the same way.

Don't really see any reason to really move it.

**Comment by Shadeykins** [comment]
https://www.youtube.com/watch?v=i7gIpuIVE3k

I'm sorry, I had to.
_don't hurt me_

**Comment by Fox-McCloud** [comment]
Fixed one Destroy loop; there's one to fix for grabs, and computer programs.

**Comment by Fox-McCloud** [comment]
That's the second one fixed.

I believe grabs will probably be next.

After that, I'm guessing I'll probably have to implement `DROPDEL`

**Comment by Fox-McCloud** [comment]
Alright, implemented `DROPDEL` which basically means "delete this item if it ever is dropped", which resolves the runtimes for a lot of items, ensured that two-handed objects don't runtime when wieldeded/unwielded, and resolved the problem with grabs runtiming, as well.

File: code/game/objects/items.dm [hunk 12]
```diff
@@ -317,6 +318,8 @@ var/global/image/fire_overlay = image("icon" = 'icons/goonstation/effects/fire.d
 	for(var/X in actions)
 		var/datum/action/A = X
 		A.Remove(user)
+	if(DROPDEL & flags)
```
Comment (tigercat2000) [inline code/game/objects/items.dm]: Why DROPDEL & flags instead of flags & DROPDEL?

Comment (Fox-McCloud) [reply code/game/objects/items.dm]: Just went with TG had; no explicit reason.

Comment (tigercat2000) [reply code/game/objects/items.dm]: Could'ya please do it the other way around? It's pretty much consistent in the code that the variable comes before the define when using stuff like == or &.

**Comment by MrStonedOne** [comment]
Our subsystems are not garbage!

File: code/game/objects/items.dm (hunk 12 above)
Comment (MrStonedOne) [reply code/game/objects/items.dm]: actually its in /tg/'s standards that the variable comes before the consistent too,, this code was out of spec.

https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#control-statements

>* All control statements comparing a variable to a number should use the formula of `thing` `operator` `number`, not the reverse (eg: `if (count <= 10)` not `if (10 >= count)`)

File: code/controllers/subsystem/garbage.dm [hunk 13]
```diff
@@ -0,0 +1,436 @@
...
+	var/tick = TICK_USAGE
+	var/ticktime = world.time
+	++delslasttick
+	++totaldels
+	var/type = D.type
+	var/refID = "\ref[D]"
+
+	del(D)
+
+	tick = (TICK_USAGE-tick+((world.time-ticktime)/world.tick_lag*100))
```
Comment (tigercat2000) [inline code/controllers/subsystem/garbage.dm]: Spessing.

File: code/controllers/subsystem/garbage.dm [hunk 14]
```diff
@@ -0,0 +1,436 @@
...
+
+	I.hard_deletes++
+	I.hard_delete_time += TICK_DELTA_TO_MS(tick)
+
+
+	if(tick > highest_del_tickusage)
+		highest_del_tickusage = tick
+	time = world.timeofday - time
+	if(!time && TICK_DELTA_TO_MS(tick) > 1)
+		time = TICK_DELTA_TO_MS(tick)/100
```
Comment (tigercat2000) [inline code/controllers/subsystem/garbage.dm]: Spacing.

File: code/controllers/subsystem/garbage.dm [hunk 15]
```diff
@@ -0,0 +1,436 @@
...
+	if(time > highest_del_time)
+		highest_del_time = time
+	if(time > 10)
+		log_game("Error: [type]([refID]) took longer than 1 second to delete (took [time/10] seconds to delete)")
```
Comment (tigercat2000) [inline code/controllers/subsystem/garbage.dm]: Spacing, spacing.

File: code/controllers/subsystem/garbage.dm [hunk 16]
```diff
@@ -0,0 +1,436 @@
...
+	var/no_hint = 0			//Number of times it's not even bother to give a qdel hint
+	var/slept_destroy = 0	//Number of times it's slept in its destroy
+
+/datum/qdel_item/New(mytype)
+	name = "[mytype]"
+
+
+// Should be treated as a replacement for the 'del' keyword.
+// Datums passed to this will be given a chance to clean up references to allow the GC to collect them.
+/proc/qdel(datum/D, force=FALSE, ...)
```
Comment (tigercat2000) [inline code/controllers/subsystem/garbage.dm]: S-spacing.

File: code/controllers/subsystem/garbage.dm [hunk 17]
```diff
@@ -0,0 +1,436 @@
...
+		for(var/varname in L)
+			if(varname == "vars")
+				continue
+			var/variable = L[varname]
+
+			if(variable == src)
+				testing("Found [src.type] \ref[src] in [D.type]'s [varname] var. [Xname]")
+
+			else if(islist(variable))
+				DoSearchVar(variable, "[Xname] -> list", recursive_limit-1)
```
Comment (tigercat2000) [inline code/controllers/subsystem/garbage.dm]: Spacing.

File: code/controllers/subsystem/garbage.dm [hunk 18]
```diff
@@ -0,0 +1,436 @@
...
+		var/normal = IS_NORMAL_LIST(X)
+		for(var/I in X)
+			if(I == src)
+				testing("Found [src.type] \ref[src] in list [Xname].")
+
+			else if(I && !isnum(I) && normal && X[I] == src)
+				testing("Found [src.type] \ref[src] in list [Xname]\[[I]\]")
+
+			else if(islist(I))
+				DoSearchVar(I, "[Xname] -> list", recursive_limit-1)
```
Comment (tigercat2000) [inline code/controllers/subsystem/garbage.dm]: Spacing.

File: code/controllers/subsystem/garbage.dm [hunk 19]
```diff
@@ -0,0 +1,436 @@
...
+		gcedlasttick = 0
+	var/cut_off_time = world.time - collection_timeout[level] //ignore entries newer then this
+	var/list/queue = queues[level]
+	var/static/lastlevel
+	var/static/count = 0
+	if(count) //runtime last run before we could do this.
+		var/c = count
+		count = 0 //so if we runtime on the Cut, we don't try again.
+		var/list/lastqueue = queues[lastlevel]
+		lastqueue.Cut(1, c+1)
```
Comment (tigercat2000) [inline code/controllers/subsystem/garbage.dm]: Spacing.

File: code/controllers/subsystem/garbage.dm [hunk 20]
```diff
@@ -0,0 +1,436 @@
...
+		count = 0 //so if we runtime on the Cut, we don't try again.
+		tobequeued.Cut(1,c+1)
+
+	for(var/ref in tobequeued)
+		count++
+		Queue(ref, GC_QUEUE_PREQUEUE+1)
+		if(MC_TICK_CHECK)
+			break
+	if(count)
+		tobequeued.Cut(1,count+1)
```
Comment (tigercat2000) [inline code/controllers/subsystem/garbage.dm]: Spacing.

File: code/controllers/subsystem/garbage.dm [hunk 21]
```diff
@@ -0,0 +1,436 @@
...
+	var/list/tobequeued = queues[GC_QUEUE_PREQUEUE]
+	var/static/count = 0
+	if(count)
+		var/c = count
+		count = 0 //so if we runtime on the Cut, we don't try again.
+		tobequeued.Cut(1,c+1)
+
+	for(var/ref in tobequeued)
+		count++
+		Queue(ref, GC_QUEUE_PREQUEUE+1)
```
Comment (tigercat2000) [inline code/controllers/subsystem/garbage.dm]: Spacing.

File: code/controllers/subsystem/garbage.dm [hunk 22]
```diff
@@ -0,0 +1,436 @@
...
+
+//If you see this proc high on the profile, what you are really seeing is the garbage collection/soft delete overhead in byond.
+//Don't attempt to optimize, not worth the effort.
+/datum/controller/subsystem/garbage/proc/HandlePreQueue()
+	var/list/tobequeued = queues[GC_QUEUE_PREQUEUE]
+	var/static/count = 0
+	if(count)
+		var/c = count
+		count = 0 //so if we runtime on the Cut, we don't try again.
+		tobequeued.Cut(1,c+1)
```
Comment (tigercat2000) [inline code/controllers/subsystem/garbage.dm]: Spacing.

File: code/controllers/subsystem/garbage.dm [hunk 23]
```diff
@@ -0,0 +1,436 @@
...
+
+/datum/controller/subsystem/garbage/fire()
+	//the fact that this resets its processing each fire (rather then resume where it left off) is intentional.
+	var/queue = GC_QUEUE_PREQUEUE
+
+	while(state == SS_RUNNING)
+		switch(queue)
+			if(GC_QUEUE_PREQUEUE)
+				HandlePreQueue()
+				queue = GC_QUEUE_PREQUEUE+1
```
Comment (tigercat2000) [inline code/controllers/subsystem/garbage.dm]: Spacing.

File: code/controllers/subsystem/garbage.dm [hunk 24]
```diff
@@ -0,0 +1,436 @@
...
+			if(GC_QUEUE_CHECK)
+				HandleQueue(GC_QUEUE_CHECK)
+				queue = GC_QUEUE_CHECK+1
```
Comment (tigercat2000) [inline code/controllers/subsystem/garbage.dm]: Spacing.

File: code/controllers/subsystem/garbage.dm [hunk 25]
```diff
@@ -0,0 +1,436 @@
...
+	if(!(delslasttick+gcedlasttick))
+		msg += "n/a|"
+	else
+		msg += "[round((gcedlasttick/(delslasttick+gcedlasttick))*100, 0.01)]%|"
+
+	msg += "TD:[totaldels]|TG:[totalgcs]|"
+	if(!(totaldels+totalgcs))
+		msg += "n/a|"
+	else
+		msg += "TGR:[round((totalgcs/(totaldels+totalgcs))*100, 0.01)]%"
```
Comment (tigercat2000) [inline code/controllers/subsystem/garbage.dm]: Spacing in operators.

As a sidenote, I wonder if we ought to ignore /tg/ and make the debug info more readable over compact- because it gets frankly kinda hard to read with everything crammed together and acronymed like this.

File: code/controllers/subsystem/garbage.dm [hunk 26]
```diff
@@ -0,0 +1,436 @@
...
+/datum/controller/subsystem/garbage/stat_entry(msg)
+	var/list/counts = list()
+	for(var/list/L in queues)
+		counts += length(L)
+	msg += "Q:[counts.Join(",")]|D:[delslasttick]|G:[gcedlasttick]|"
+	msg += "GR:"
+	if(!(delslasttick+gcedlasttick))
+		msg += "n/a|"
+	else
+		msg += "[round((gcedlasttick/(delslasttick+gcedlasttick))*100, 0.01)]%|"
```
Comment (tigercat2000) [inline code/controllers/subsystem/garbage.dm]: Spacing in operators.

File: code/modules/mob/mob_grab.dm [hunk 27]
```diff
@@ -434,7 +434,8 @@
 	return EAT_TIME_FAT //if it doesn't fit into the above, it's probably a fat guy, take EAT_TIME_FAT to do it
 
 /obj/item/weapon/grab/dropped()
-	qdel(src)
+	if(!QDELETED(src))
```
[the remark this answers is not in the document]
Comment (Fox-McCloud) [reply code/modules/mob/mob_grab.dm]: You think it would be appropriate for grabs? I instantly thought of this, but wasn't sure if appropriate or not.

[the remark this answers is not in the document]
Comment (tigercat2000) [reply code/modules/mob/mob_grab.dm]: All of their cleanup happens in /Destroy so I don't see why not (as long as it actually works lol)
